timelock-server/src/suiteTest/java/com/palantir/atlasdb/timelock/MultiLeaderMultiNodePaxosTimeLockIntegrationTest.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.timelock;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import java.util.Set;
+import java.util.UUID;
+import java.util.stream.Stream;
+
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ImmutableSetMultimap;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Multimaps;
+import com.google.common.collect.SetMultimap;
+import com.palantir.atlasdb.timelock.suite.MultiLeaderPaxosSuite;
+import com.palantir.atlasdb.timelock.util.ParameterInjector;
+import com.palantir.common.streams.KeyedStream;
+import com.palantir.conjure.java.api.errors.QosException;
+
+@RunWith(Parameterized.class)
+public class MultiLeaderMultiNodePaxosTimeLockIntegrationTest {
+
+    @ClassRule
+    public static ParameterInjector<TestableTimelockCluster> injector =
+            ParameterInjector.withFallBackConfiguration(() -> MultiLeaderPaxosSuite.MULTI_LEADER_PAXOS);
+
+    @Parameterized.Parameter
+    public TestableTimelockCluster cluster;
+
+    @Parameterized.Parameters(name = "{0}")
+    public static Iterable<TestableTimelockCluster> params() {
+        return injector.getParameter();
+    }
+
+
+    @Test
+    public void eachNamespaceGetsAssignedDifferentLeaders() {
+        SetMultimap<TestableTimelockServer, String> allocations = allocateRandomNamespacesToAllNodes(3);
+
+        Set<String> allNamespaces = ImmutableSet.copyOf(allocations.values());
+
+        allNamespaces.stream()
+                .map(cluster::client)
+                .map(NamespacedClients::throughWireMockProxy)
+                .forEach(NamespacedClients::getFreshTimestamp);
+
+        SetMultimap<TestableTimelockServer, String> namespacesByLeader =
+                ImmutableSetMultimap.copyOf(cluster.currentLeaders(allNamespaces)).inverse();
+
+        assertThat(namespacesByLeader)
+                .isEqualTo(allocations);
+
+        cluster.servers().forEach(TestableTimelockServer::allowAllNamespaces);
```

Comment:
Discussed offline: The reason this test works is because in the setup, if we have assigned namespace `NAMESPACE` to node `A`, then only `A` knows about `NAMESPACE`. So only `A` will propose in the Paxos sequence for `NAMESPACE`, and thus `A` should become the leader. Also, before we `allowAllNamespaces` the other nodes will just reply with 503s that are in front of the direct proxy, so it's not possible that there will be another election in the other namespaces causing `cluster.currentLeaders(...)` to change.

This is pretty clever! Maybe worth a comment somewhere.

Reply:
with some of my other work, we can probably use the takeover primitive instead to more reliably deal with this.